File: 94. Binary Tree Inorder Traversal/solution_2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

 class Solution 
 {
     public List<Integer> inorderTraversal(TreeNode root) 
     {
         List<Integer> values = new ArrayList<>();
         List<TreeNode> stack = new ArrayList<>();
         
         TreeNode node = root;
         
         while (node != null || stack.isEmpty() == false)
         {
             while (node != null)
             {
                 stack.add(node);
                 node = node.left;
             }
             
             node = stack.getLast();
             stack.removeLast();
             
             values.add(node.val);
             
             node = node.right;
         }
         
         return values;
     }
 }

File: 10-99/94. Binary Tree Inorder Traversal/solution_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution 
{
    public:
        vector<int> inorderTraversal(TreeNode* root) 
        {
            vector<int> values;
            vector<TreeNode*> stack;

            TreeNode* node = root;

            while(node != nullptr || stack.empty() == false)
            {
                while (node != nullptr)
                {
                    stack.push_back(node);
                    node = node->left;
                }

                node = stack.back();
                stack.pop_back();

                values.push_back(node->val);

                node = node->right;
            }

            return values;
        }
};